### PR TITLE
Feature: add a linear fit function by least square method

### LINKED ISSUE #3815: Feature: need a math function to do linear fit
### Background

I try figure it out whether SCF is trapped in a local min #3755. To this end, I need some math function. One of them is `Linear_Fit`.

### Describe the solution you'd like

I will add it in `Base` class.

### Task list only for developers

- [ ] Notice possible changes of behavior
- [ ] Explain the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi

### Notice Possible Changes of Behavior (Reminder only for developers)

_No response_

### Notice any changes of core modules (Reminder only for developers)

_No response_

### Notice Possible Changes of Core Modules (Reminder only for developers)

_No response_

### Additional Context

_No response_

### Task list for Issue attackers (only for developers)

- [ ] Review and understand the proposed feature and its importance.
- [ ] Research on the existing solutions and relevant research articles/resources.
- [ ] Discuss with the team to evaluate the feasibility of implementing the feature.
- [ ] Create a design document outlining the proposed solution and implementation details.
- [ ] Get feedback from the team on the design document.
- [ ] Develop the feature following the agreed design.
- [ ] Write unit tests and integration tests for the feature.
- [ ] Update the documentation to include the new feature.
- [ ] Perform code review and address any issues.
- [ ] Merge the feature into the main branch.
- [ ] Monitor for any issues or bugs reported by users after the feature is released.
- [ ] Address any issues or bugs reported by users and continuously improve the feature.

### PR DESCRIPTION
**Pull request by WHUweiqingzhou**
Fix #3815.

Add a linear fit function by least square method.

### The List of Changes
1. add `math_tools.h` and `math_tools.cpp` in `Base` class to store fundamental mathematical function.
2. add `ModuleBase::linear_fit()` to perform linear fit by least square method.
3. add correspond UnitTests to protect it.

### CODE REVIEW

File: source/module_base/CMakeLists.txt [hunk 1]
```diff
@@ -31,6 +31,7 @@ add_library(
     math_bspline.cpp
     math_chebyshev.cpp
     mathzone_add1.cpp
+    math_tools.cpp
```
Comment (dyzheng) [inline source/module_base/CMakeLists.txt:34]: Makefile.Objects

**Comment by kirk0830** [comment]
I am curious about the idea how you use least square regression to tackle the SCF trapping-local-minima problem, could you explain a little bit?

File: source/module_base/math_tools.cpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+std::pair<double, double> linear_fit(const std::vector<double>& x, const std::vector<double>& y)
+{
+    assert(x.size() == y.size()); // the size of x and y should be the same
+    assert(x.size() > 0); // the size of x should be larger than 0
+    // get the size
+    const int n = x.size();
+
+  // calculate the sum of x, y, xy, xx
+    double x_sum = 0.0, y_sum = 0.0, x2_sum = 0.0, xy_sum = 0.0;
```
Comment (mohanchen) [inline source/module_base/math_tools.cpp:15]: Initialize a variable for each line.

File: source/module_base/math_tools.cpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace ModuleBase
+{
+
+std::pair<double, double> linear_fit(const std::vector<double>& x, const std::vector<double>& y)
+{
+    assert(x.size() == y.size()); // the size of x and y should be the same
+    assert(x.size() > 0); // the size of x should be larger than 0
+    // get the size
+    const int n = x.size();
```
Comment (mohanchen) [inline source/module_base/math_tools.cpp:12]: add a check point here: 'n' should not be zero

File: source/module_base/math_tools.cpp [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        y_sum += y[i];
+        x2_sum += x[i] * x[i];
+        xy_sum += x[i] * y[i];
+    }
+
+    double x_mean = x_sum / n;
+    double y_mean = y_sum / n;
+
+    //
+    double slope = (n * xy_sum - x_sum * y_sum) / (n * x2_sum - x_sum * x_sum);
```
Comment (mohanchen) [inline source/module_base/math_tools.cpp:28]: check: will the (n * x2_sum - x_sum * x_sum) denominator  be zero?

**Comment by WHUweiqingzhou** [comment]
I close this PR temporally.

**Pull request closed without merging**